### ISSUE TITLE
Compiler emits warning `unreachable pattern` when invoking macro in pattern

### ISSUE THREAD

**Issue by lidin** [rust-lang/rust#83211]
I tried this code:

```rust
macro_rules! choose {
    ($choice:literal) => { $choice };
}

macro_rules! choose_reachable {
    (true) => { true };
    (false) => { false };
}

const _UNREACHABLE_PATTERN: () = match false {
    choose!(true) | choose!(false) => (),
};

const _REACHABLE_PATTERN: () = match false {
    choose_reachable!(true) | choose_reachable!(false) => (),
};

fn main() {}
```

I expected to see this happen: 
No warnings or errors.

Instead, this happened: 
Compiler emits warning `unreachable pattern`. This does not happen in nightly, so I don't know if this has been fixed already.

```
warning: unreachable pattern
  --> src/main.rs:3:9
   |
3  |         $choice
   |         ^^^^^^^
...
17 |     choose!(true) | choose!(false) => (),
   |                     -------------- in this macro invocation
   |
   = note: `#[warn(unreachable_patterns)]` on by default
   = note: this warning originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

warning: 1 warning emitted
```

### Meta

`rustc --version --verbose`:
```
rustc 1.50.0 (cb75ad5db 2021-02-10)
binary: rustc
commit-hash: cb75ad5db02783e8b0222fee363c5f63f7e2cf5b
commit-date: 2021-02-10
host: x86_64-apple-darwin
release: 1.50.0
```

<details><summary>Backtrace</summary>
<p>

No backtrace generated.

</p>
</details>

**Comment by jonas-schievink** [comment]
This is fixed on nightly
